### PR TITLE
Fixed #35667 -- Improved deprecation warning handling by replacing stacklevel with skip_file_prefixes.

### PR DESCRIPTION
**Pull request by nessita**
#### Trac ticket number

ticket-35667

### CODE REVIEW

File: django/utils/deprecation.py [hunk 1]
```diff
@@ -20,6 +19,20 @@ class RemovedInDjango61Warning(PendingDeprecationWarning):
 RemovedAfterNextVersionWarning = RemovedInDjango61Warning
 
 
+def emit_warning(msg, category=None, stacklevel=2):
+    # TODO: Remove stacklevel when dropping support for PY311.
+    if PY312:
+        caller_frame = inspect.stack()[1]
+        caller_file = caller_frame.frame.f_globals.get("__file__", None)
+        kwargs = {"skip_file_prefixes": (os.path.dirname(caller_file),)}
```
Comment (charettes) [inline django/utils/deprecation.py:27]: I don't think that's what we want and I suspect some tests will fail. What this does is retrieve the package of the caller and pass it as `skip_file_prefixes` but this doesn't account for user code indirectly calling into a nested package emitting a warning.

As an example, think of a method in `django.db.models.sql.query` that emit a warnings when called with a particular signature that is used by a public interface in `django.db.models`. When user code calls into the public interface exposed by the public interface we want the deprecation warning to be raised at the user call site and not at `django.db.models` _proxy_ call site.

The code above would make it so `skip_file_prefixes = (os.path.dirname(django.db.models.sql.__file__),)` which bakes the assumption that we'll always emit warnings one package away from publicly exposed interfaces.

To put it simply I think we should simply set `skip_file_prefixes=(os.path.dirname(django.__file__),)` systmetically on `PY312` once we've confirmed that we don't make use of the deprecated code internally and do away with the `stacklevel` arithmetic.

Comment (nessita) [reply django/utils/deprecation.py:27]: Thank you @charettes for this detailed feedback. :star2:
It's in my ToDo to review in depth and progress/fix this. I will close this PR and continue future conversations/proposals in the original PR.
